From 486c417dfea3c61a601d0b3792ba967cc9216367 Mon Sep 17 00:00:00 2001
From: Davor Joja <davorjoja@logicbricks.com>
Date: Wed, 5 Nov 2014 16:51:00 +0100
Subject: [PATCH 0446/1566] gpu: drm: xylon: Fixed NULL pointer dereference

Call drm_fb_helper_prepare() function for proper struct drm_fb_helper
initialization and remove manual drm_fb_helper_funcs pointer initialization.

Error log:
Unable to handle kernel NULL pointer dereference at virtual address 000001d0
pgd = 40004000
[000001d0] *pgd=00000000
Internal error: Oops - BUG: 5 [#1] PREEMPT SMP ARM
Modules linked in:
CPU: 0 PID: 6 Comm: kworker/u4:0 Not tainted 3.17.0
PC is at drm_fb_helper_single_add_all_connectors+0x10/0xa8
LR is at xylon_drm_fbdev_init+0x7c/0xfc

Signed-off-by: Davor Joja <davorjoja@logicbricks.com>
Tested-by: Christian Kohn <christian.kohn@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 drivers/gpu/drm/xylon/xylon_fbdev.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/xylon/xylon_fbdev.c b/drivers/gpu/drm/xylon/xylon_fbdev.c
index 0d848c7..85c3bb9 100644
--- a/drivers/gpu/drm/xylon/xylon_fbdev.c
+++ b/drivers/gpu/drm/xylon/xylon_fbdev.c
@@ -127,7 +127,7 @@ err_fb_alloc:
 	return ret;
 }
 
-static struct drm_fb_helper_funcs xylon_drm_fbdev_helper_funcs = {
+static const struct drm_fb_helper_funcs xylon_drm_fbdev_helper_funcs = {
 	.fb_probe = xylon_drm_fbdev_create,
 };
 
@@ -146,9 +146,10 @@ xylon_drm_fbdev_init(struct drm_device *dev,
 		return ERR_PTR(-ENOMEM);
 	}
 
-	fbdev->fb_helper.funcs = &xylon_drm_fbdev_helper_funcs;
 	helper = &fbdev->fb_helper;
 
+	drm_fb_helper_prepare(dev, helper, &xylon_drm_fbdev_helper_funcs);
+
 	ret = drm_fb_helper_init(dev, helper, num_crtc, max_conn_count);
 	if (ret < 0) {
 		DRM_ERROR("failed fb init\n");
-- 
1.7.5.4

